### PR TITLE
Dynamic styling

### PR DESCRIPTION
**Pull request by Nevon**
This is needed for merchant styling in KCO. This is exposing a `customize` prop on certain components, which then implements the dynamic styling from the ui-css-components.

Components:

* [x] Link
* [x] Switch
* [x] Button
* [x] Field

### CODE REVIEW

File: components/Loader/index.jsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+Loader.propTypes = {
+  className: PropTypes.string,
+  color: PropTypes.oneOfType([
+    PropTypes.oneOf(Object.keys(colors)),
+    PropTypes.array
+  ]),
+  size: PropTypes.oneOf(Object.keys(sizes)),
+  styles: PropTypes.object
+}
```
Comment (joaomilho) [inline components/Loader/index.jsx]: The only cool thing is that the blame on this thing will point to you now. 😈

File: components/Switch.jsx [hunk 2]
```diff
@@ -106,3 +117,70 @@ Switch.propTypes = {
...
+    onClick={onClick}
+    onMouseDown={onMouseDown}
+    onMouseUp={onMouseUp}
+    {...remainingProps}>
+    <div
+      className={childClassNames.bullet}
+      style={{
+        backgroundColor: backgroundColor,
+        borderColor: backgroundColor
+      }}></div>
```
Comment (joaomilho) [inline components/Switch.jsx]: Is there any good reason for this not to be a regular checkbox?

**Comment by joaomilho** [comment]
Just one question, why don't we do customizations as context? Seems to be the right usage for it.

**Comment by xaviervia** [comment]
@joaomilho I believe context is an anti pattern. That's why initially the implementation of the Entanglement was modified not to use it. Context is akin to global variables and I'm extremely wary of sending the message that it is OK to use that.

I am aware that passing down this information through the tree is going to be cumbersome, but the exact same argument can be made for passing down values of inputs (after all, they could be set on a generic key-value object in context). I think all the arguments against global variables hold here: prone to errors, hard to debug, inviting to bad practices. This library is going to be used by a lot of people new to React and it's going to be used as reference of legitimate practices: I am all the more wary about context because of that.

Again, the only reason I agree on context's usage in the Entanglement is because having context is the only reason the Scatter as a component API that renders locally or remotely makes sense. Otherwise it should always be a function (no need for a component). In other words, I agree to context in that case because the whole `<Scatter>` Entanglement syntax is sugar anyway, and can be safely dropped for a direct calls to the adapter if one feels so inclined.

File: components/Switch.jsx (hunk 2 above)
Comment (xaviervia) [inline components/Switch.jsx]: No, not at all. The CSS components don't currently support that, but we should most definitely implement it as a regular one.

**Comment by pirelenito** [comment]
Hi there! Coming really late to the discussion...

Sorry @xaviervia but I have to disagree with you. I understand your position on trying not to push `context` as a "silver bullet" to any team using this library as a reference, but ignoring its use-case for this particular functionality might be more problematic.

Dynamic styling/themes is a completely different problem than passing down regular information down the tree, the biggest difference being that we want the same information available all the time for a component of a given type. This means that all buttons must look the same, so we will always have these customization properties being passed down, which I think is a even bigger anti-pattern, specially if we want consistency in the interface.

Again, I think we should think about a solution using `context` and maybe take this time as an opportunity to discuss styling in general.

;)

**Comment by joaomilho** [comment]
We had a long discussion over Slack not long ago, but we got to the same point. Can we sum the discussion like this, or am I making a straw man of the opposition:

- We're not all adults;
- Since we're not adults, people will forget to pass the props and it will defeat the purpose of merchant styling anyway. Not only context is much easier, but it avoids any inconsistency in styling by design;

**Comment by pirelenito** [comment]
Here is a follow up: https://github.com/klarna/ui-react-components/pull/66

**Pull request merged**